src/Symfony/Component/Serializer/Normalizer/AbstractNormalizer.php

```diff
@@ -32,20 +32,80 @@ abstract class AbstractNormalizer implements NormalizerInterface, DenormalizerIn
     use ObjectToPopulateTrait;
     use SerializerAwareTrait;
 
-    const CIRCULAR_REFERENCE_LIMIT = 'circular_reference_limit';
-    const OBJECT_TO_POPULATE = 'object_to_populate';
-    const GROUPS = 'groups';
-    const ATTRIBUTES = 'attributes';
-    const ALLOW_EXTRA_ATTRIBUTES = 'allow_extra_attributes';
+    /* constants to configure the context */
+    /**
+     * How many loops of circular reference to allow while normalizing.
+     *
+     * The default value of 1 means that when we encounter the same object a
+     * second time, we consider that a circular reference.
+     *
+     * You can raise this value for special cases, e.g. in combination with the
+     * max depth setting of the object normalizer.
+     */
+    public const CIRCULAR_REFERENCE_LIMIT = 'circular_reference_limit';
+
+    /**
+     * Instead of creating a new instance of an object, update the specified object.
+     *
+     * If you have a nested structure, child objects will be overwritten with
+     * new instances unless you set DEEP_OBJECT_TO_POPULATE to true.
+     */
+    public const OBJECT_TO_POPULATE = 'object_to_populate';
+
+    /**
+     * Only (de)normalize attributes that are in the specified groups.
+     */
+    public const GROUPS = 'groups';
+
+    /**
+     * Limit (de)normalize to the specified names.
+     *
+     * For nested structures, this list needs to reflect the object tree.
+     */
+    public const ATTRIBUTES = 'attributes';
+
+    /**
+     * If ATTRIBUTES are specified, and the source has fields that are not part of that list,
+     * either ignore those attributes (true) or throw an ExtraAttributesException (false).
+     */
+    public const ALLOW_EXTRA_ATTRIBUTES = 'allow_extra_attributes';
+
     const DEFAULT_CONSTRUCTOR_ARGUMENTS = 'default_constructor_arguments';
-    const CALLBACKS = 'callbacks';
-    const CIRCULAR_REFERENCE_HANDLER = 'circular_reference_handler';
-    const IGNORED_ATTRIBUTES = 'ignored_attributes';
 
     /**
-     * @internal
```

Comment:
Why this removal? IIRC, we’re in the process of deprecating this feature in favor of custom normalizers and decoration.

Reply:
we are now on php 7.1, i changed accessibility to the constant to `protected`.

but i can revert that change if you want, or make it protected while keeping the annotation.

Reply:
Yes please revert, since we allow other library to extends this normalizer removing this annotation will maybe induce a BC break later.

Reply:
i understood that extending classes could use it (AbstractObjectNormalizer actually does).

so should i make it `protected` but keep the `@internal` annotation?

Reply:
Don't mind having it protected yes, but let's keep `@internal` annotation